### ISSUE TITLE
Type system for the meta language

### ISSUE THREAD

**Issue by SkySkimmer** [Andromedans/andromeda#229]
I think we want to get one. What should it look like?

**Comment by andrejbauer** [comment]
Ha, that is a good question, find @matijapretnar and pick his brain. My first guess would be that it should look like ML with an abstract type `jdg` of judgements (maybe call it `tt`?)

**Comment by andrejbauer** [comment]
Maybe we can have `coerce` first, as that would be quite useful?

**Comment by matijapretnar** [comment]
This should be relatively easy to do, so: an ML type system together with a type of judgements. Do you want an effect system as well?

**Comment by andrejbauer** [comment]
I think we can skip the effect system as that will significantly simplify everything, and it's not really cruicial. There are a couple of quirks, however, I hope they won't create a problem. For instance an application `e₁ e₂` may mean either the meta-language application or the object-language application, depending on the type of `e₁`. There are probably a couple of other such dynamic features in the current implementation, some of which ought to be easily removable.

**Comment by SkySkimmer** [comment]
What would an effect system give us?

Basically there are 2 rules
```
|- e1 : t1 -> t2
|- e2 : t1
________________
|- e1 e2 : t2
```
and 
```
|- e1 : jdg
|- e2 : jdg
______________
|- e1 e2 : jdg
```

**Comment by matijapretnar** [comment]
The effect system would allow you to track which operations are called. But I guess that is not too useful at this point, as all the operations are the ones issued by the kernel, and each computation is bound to call them. So cancel that.

As for the two typing rules, I don't know the best solution. For example, what is the type of
`fun x => fun y => x y`? Is it `jdg -> jdg -> jdg` or `(t1 -> t2) -> t1 -> t2`?

Maybe we can mirror the evaluation in the inference. First infer whether `e1` and `e2` are of type `jdg`. If that works, output the type to be `jdg`, otherwise try with function types. What worries me is that we may not have enough information at that point yet. Say that we have an application of the above function `(fun x => fun y => x y) e1 e2`. When we're inferring the type of `x y`, we haven't looked at `e1` and `e2` yet. Maybe if we rewrote inference in CPS, we could try and finish inferring the whole computation by assuming all types are `jdg`, and then backtrack if that fails.

**Comment by SkySkimmer** [comment]
How about subtyping `jdg :> jdg -> jdg`? That's pretty much how we use it after all.

**Comment by andrejbauer** [comment]
By the way, before we lay on massively more features, I would like to have a phase of "cleaning" and, more importantly, documenting the source code.

**Comment by andrejbauer** [comment]
You could solve a lot of the dilemmas by not insisting on the ability to infer types 100% of the time, so perhaps typing annotations would be needed in certain situations to help resolve ambiguities.

**Comment by matijapretnar** [comment]
Yes, annotations could resolve ambiguities. We would still need some modifications of HM to sort out inference. For example, when you annotate `fun x => fun y => x y : jdg -> jdg -> jdg`, the top annotation still needs to trickle down to the application `x y`.

One possible solution (I did not really think it through, but still…) is to have other kind of constraints beside equations. Then, when you see `e1 e2` and infer `e1 : t1` & `e2 : t2`, you don't just come up with a fresh `t` and add a constraint `t1 = t2 -> t`, but add a special constraint `app_constraint(t1, t2, t)` that you resolve later when you have more information. For example, you can resolve all the equations and see whether `t1` is `jdg` or some arrow type.

**Comment by andrejbauer** [comment]
I was rather thinking of annotation in binders, i.e., `fun (x : A) => ...` instead of `fun x => ...`, isn't that a simple solution that would work? In case you want a polymorphic function you just pick a parameter, i.e., `fun (x : List(α * int)) => ...` would mean that `x` is a certain polymorphic list.

As a general strategy I'd prefer a simple kind of type theory even at the expense of user comfort. Anything complicated is not worth implementing at this point, it's too likely that we'll want to change it later.

**Comment by andrejbauer** [comment]
Closed by #321
